sys/amd64/amd64/patch_machdep.c

```diff
@@ -0,0 +1,150 @@
+#include <sys/types.h>
+#include <sys/param.h>
+#include <sys/systm.h>
+#include <sys/domain.h>
+#include <sys/domainset.h>
+#include <sys/kernel.h>
+#include <sys/lock.h>
+#include <sys/malloc.h>
+#include <sys/mutex.h>
+#include <sys/queue.h>
+#include <sys/vmem.h>
+#include <sys/zcond.h>
+
+#include <vm/vm.h>
+#include <vm/pmap.h>
+#include <vm/uma.h>
+#include <vm/vm_domainset.h>
+#include <vm/vm_extern.h>
+#include <vm/vm_page.h>
+#include <vm/vm_pagequeue.h>
+
+#include <machine/cpufunc.h>
+#include <machine/pmap.h>
+#include <machine/patch.h>
+
+static struct pmap patch_pmap;
+static vm_offset_t patch_va;
+static pt_entry_t *patch_pte;
+
+static pt_entry_t *
+patch_init_pte(void)
+{
+	vm_page_t dummy_page;
+	int domain;
+	pml5_entry_t *pml5e;
+	pml4_entry_t *pml4e;
+	pdp_entry_t *pdpe;
+	pd_entry_t *pde;
+	pt_entry_t *pte;
+	vm_pindex_t pml5_idx, pml4_idx, pdp_idx, pd_idx;
+	vm_paddr_t mphys;
+	extern int la57;
+	bool is_la57;
+
+	domain = PCPU_GET(domain);
+	vmem_alloc(VM_DOMAIN(domain)->vmd_kernel_nofree_arena, PAGE_SIZE,
+	    M_BESTFIT | M_WAITOK, &patch_va);
+	dummy_page = vm_page_alloc_noobj_domain(domain,
+	    VM_ALLOC_WIRED | VM_ALLOC_NOFREE);
+	pmap_enter(&patch_pmap, patch_va, dummy_page, VM_PROT_WRITE,
+	    PMAP_ENTER_WIRED, 0);
+
+	is_la57 = false;
+	if (patch_pmap.pm_type == PT_X86) {
+		is_la57 = la57;
+	}
+
+	pml4_idx = pmap_pml4e_index(patch_va);
+	if (is_la57) {
+		pml5_idx = pmap_pml5e_index(patch_va);
+		pml5e = &patch_pmap.pm_pmltopu[pml5_idx];
+		KASSERT(*pml5e != 0,
+		    ("%s: va %#jx pml5e == 0", __func__, patch_va));
+		mphys = *pml5e & PG_FRAME;
+
+		pml4e = (pml4_entry_t *)PHYS_TO_DMAP(mphys);
+		pml4e = &pml4e[pml4_idx];
+	} else {
+		pml4e = &patch_pmap.pm_pmltop[pml4_idx];
+	}
+
+	KASSERT(*pml4e != 0,
+	    ("%s: va %#jx pml4e == 0", __func__, patch_va));
+	mphys = *pml4e & PG_FRAME;
+
+	pdpe = (pdp_entry_t *)PHYS_TO_DMAP(mphys);
+	pdp_idx = pmap_pdpe_index(patch_va);
+	pdpe += pdp_idx;
+	KASSERT(*pdpe != 0,
+	    ("%s: va %#jx pdpe == 0", __func__, patch_va));
+	mphys = *pdpe & PG_FRAME;
+
+	pde = (pd_entry_t *)PHYS_TO_DMAP(mphys);
+	pd_idx = pmap_pde_index(patch_va);
+	pde += pd_idx;
+	KASSERT(*pde != 0, ("%s: va %#jx pde == 0", __func__, patch_va));
+	mphys = *pde & PG_FRAME;
+
+	pte = (pt_entry_t *)PHYS_TO_DMAP(mphys);
+	pte += pmap_pte_index(patch_va);
+
+	return (pte);
+}
+
+static void
+patch_pmap_init(const void *unused)
+{
+	vm_offset_t kern_start, kern_end;
+
+	kern_start = virtual_avail;
+	kern_end = kernel_vm_end;
```

Comment:
Moram priznat da mi je ovo promaklo u prvoj iteraciji, a vidim da je mark to ulovio.
Koliko sam ubrao, ovo si vjerojatno kopirao iz pmap init dijela, što u tom kontekstu radi jer u tom stadiju boota aktivan samo jedan procesor.

Međutim `virtual_avail` se dosta mijenja i ovo nije ispravno.
Najispravnija stvar je iščitati i kopirati VA range iz `kernel_linker_file` varijable (`->address` i `->size`), jer linker može relocateati kernel na neku drugu adresu tijekom boota.